core/trino-spi/src/test/java/io/trino/spi/TestSpiBackwardCompatibility.java

```diff
@@ -0,0 +1,138 @@
+        assertThat(getCurrentSpi()).containsAll(difference(getPreviousSpi(), backwardIncompatibleChanges));
+    }
+
+    private static String getPreviousSpiVersion()
+    {
+        try {
+            URL resource = TestSpiBackwardCompatibility.class.getClassLoader().getResource("io/trino/spi/previous-trino-spi-version.txt");
+            requireNonNull(resource, "version resource not found");
+            try (InputStream inputStream = resource.openStream();
+                    BufferedReader reader = new BufferedReader(new InputStreamReader(inputStream, UTF_8))) {
+                String spiVersion = reader.readLine();
+                if (spiVersion == null || spiVersion.isBlank() || reader.readLine() != null) {
+                    throw new IllegalStateException("Malformed version resource");
+                }
+                return spiVersion.strip();
+            }
+        }
+        catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static Set<String> getCurrentSpi()
+            throws IOException
+    {
+        return getSpiEntities(getSystemClassLoader(), true);
+    }
+
+    private static Set<String> getPreviousSpi()
+            throws Exception
+    {
+        try (Stream<Path> list = Files.list(Path.of("target", "released-artifacts"))) {
+            URL[] jars = list.map(TestSpiBackwardCompatibility::getUrl)
+                    .toArray(URL[]::new);
+            try (URLClassLoader urlClassLoader = new URLClassLoader(jars, getPlatformClassLoader())) {
+                return getSpiEntities(urlClassLoader, false);
+            }
+        }
+    }
+
+    private static URL getUrl(Path path)
+    {
+        try {
+            return path.toUri().toURL();
+        }
+        catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static Set<String> getSpiEntities(ClassLoader classLoader, boolean includeDeprecated)
+            throws IOException
+    {
+        ImmutableSet.Builder<String> entities = ImmutableSet.builder();
+        for (ClassInfo classInfo : ClassPath.from(classLoader).getTopLevelClassesRecursive("io.trino.spi")) {
+            Class<?> clazz = classInfo.load();
+            if (!includeDeprecated && clazz.isAnnotationPresent(Deprecated.class)) {
+                continue;
+            }
+            for (Constructor<?> constructor : clazz.getConstructors()) {
+                if (includeDeprecated || !constructor.isAnnotationPresent(Deprecated.class)) {
+                    entities.add("Constructor: " + constructor.toGenericString());
+                }
+            }
+            for (Method method : clazz.getMethods()) {
+                if (includeDeprecated || !method.isAnnotationPresent(Deprecated.class)) {
+                    entities.add("Method: " + method.toGenericString());
```

Comment:
This is recursive (higher up in class hierarchy), maybe worth commenting about this.

Also, you returning Object-defined methods

```
public final void java.lang.Object.wait(long,int) throws java.lang.InterruptedException
public final void java.lang.Object.wait() throws java.lang.InterruptedException
public final native void java.lang.Object.wait(long) throws java.lang.InterruptedException
public boolean java.lang.Object.equals(java.lang.Object)
public java.lang.String java.lang.Object.toString()
public native int java.lang.Object.hashCode()
public final native java.lang.Class<?> java.lang.Object.getClass()
public final native void java.lang.Object.notify()
public final native void java.lang.Object.notifyAll()
```

they will not cause a change (unless last SPI classes is removed), but still maybe worth documenting?

Reply:
I changed it to return only declared entities and skip object and enum methods. Thanks.

Previously, used `getMethods` and `getFields` as it returns only public things. However it is easier public things than hierarchical ones (or maybe it as same level of complexity...)